peps/pep-0728.rst

```diff
@@ -0,0 +1,633 @@
+
+Rationale
+=========
+
+A type that allows extra fields of type ``str`` on a TypedDict can be loosely
+described as the intersection between the TypedDict and ``Mapping[str, str]``.
+
+`Index Signatures <https://www.typescriptlang.org/docs/handbook/2/objects.html#index-signatures>`__
+in TypeScript achieve this:
+
+.. code-block:: typescript
+
+    type Foo = {
+        a: string
+        [key: string]: string
+    }
+
+This proposal aims to support a similar feature without introducing general
+intersection of types or syntax changes.
+
+We propose that we give the dunder attribute ``__extra__`` a special meaning:
+When it is defined on a TypedDict type, extra fields are allowed, and their
+types should be compatible with the type of ``__extra__``. Different from index
+signatures, the types of known fields don't need to be consistent with the type
+of ``__extra__``.
+
+There are some advantages to this approach:
+
+- Inheritance works naturally. ``__extra__`` defined on a TypedDict will also
+  be available to its subclasses.
+
+- We can build on top of
+  :pep:`the type consistency rules defined in PEP 589 <589#type-consistency>`.
+  ``__extra__`` can be treated as a pseudo-key in terms of type consistency.
+
+- There is no need to introduce a syntax to specify the type of the extra
+  fields.
+
+- We can precisely type the extra fields without making ``__extra__`` the union
+  of known fields.
+
+Specification
+=============
+
+This specification is structured to parallel :pep:`589` to highlight changes to
+the original TypedDict specification.
+
+Extra fields are treated as non-required fields having the same type of
+``__extra__`` whose keys are allowed when determining
+:pep:`supported and unsupported operations <589>`.
+
+Using TypedDict Types
+---------------------
+
+For a TypedDict type that has the ``__extra__`` field, during construction, the
+value type of each unknown field is expected to be non-required and compatible
+with the type of ``__extra__``. For example::
+
+    class Movie(TypedDict):
+        name: str
+        __extra__: bool
+    
+    a: Movie = {"name": "Blade Runner", "novel_adaptation": True}  # OK
+    b: Movie = {
+        "name": "Blade Runner",
+        "year": 1982,  # Not OK. 'int' is incompatible with 'bool'
+    }  
+
+In this example, ``__extra__: bool`` does not mean that ``Movie`` has a required
+string key "__extra__" whose value type is ``bool``. Instead, it specifies that
+fields other than "name" have a value type of ``bool`` and are non-required.
+
+The alternative inline syntax is also supported::
+
+    Movie = TypedDict("Movie", {"name": str, "__extra__": bool})
+
+Accessing extra fields is allowed. Type checkers must infer its value type from
+the type of ``__extra__``::
+
+    def f(movie: Movie, key: str) -> None:
+        reveal_type(movie["name"])  # Revealed type is 'str'
+        reveal_type(movie["year"])  # Revealed type is 'int'
+
+
+Interaction with Totality
```

Comment:
We'll also need to specify the interaction with PEP 705. It seems to me that `ReadOnly` should be allowed on `__extra__`, with the obvious meaning: all the extra fields are read-only.